src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,113 @@
+            new Translation2d(Rx, Ry),
+            new Translation2d(-Rx, Ry),
+            new Translation2d(Rx, -Ry),
+            new Translation2d(-Rx, -Ry)
+    );
+    private final Timer timer = new Timer();
+    private double currentTime;
+    private double lastTime = 0;
+    private boolean fieldOriented;
+
+    private static SwerveDrive INSTANCE = null;
+
+    /**
+     * Constructor.
+     *
+     * @param fieldOriented whether the robot is field oriented.
+     */
+    private SwerveDrive(boolean fieldOriented) {
+        this.fieldOriented = fieldOriented;
+
+        timer.start();
+        configModules();
+    }
+
+    public static SwerveDrive getInstance(){
+        if(INSTANCE == null)
+            INSTANCE = new SwerveDrive(true);
+        return INSTANCE;
+    }
+
+    /**
+     * Sets the states for the swerve drive.
+     *
+     * @param vx       is the velocity in the x direction. [m/s]
+     * @param vy       is the velocity in the y direction. [m/s]
+     * @param rotation is the rotation velocity. [rad/s]
+     */
+    public void holonomicDrive(double vx, double vy, double rotation) {
+        ChassisSpeeds speeds = fieldOriented ?
+                ChassisSpeeds.fromFieldRelativeSpeeds(
+                        vx, vy, rotation, Rotation2d.fromDegrees(Robot.navx.getYaw())
+                ) :
+                new ChassisSpeeds(vx, vy, rotation);
+
+        SwerveModuleState[] states = kinematics.toSwerveModuleStates(speeds);
+        SwerveDriveKinematics.normalizeWheelSpeeds(states, MAX_VELOCITY);
+
+        for (int i = 0; i < 4; i++) {
+            states[i] = SwerveModuleState.optimize(states[i], Rotation2d.fromDegrees(modules[i].getAngle()));
+
+            modules[i].setAngle(states[i].angle.getDegrees());
+            modules[i].setVelocity(states[i].speedMetersPerSecond, currentTime - lastTime);
+        }
+    }
+
+    /**
+     * Sets whether the robot is field oriented.
+     *
+     * @param fieldOriented is whether the robot is field oriented.
+     */
+    public void setFieldOriented(boolean fieldOriented) {
+        this.fieldOriented = fieldOriented;
+    }
+
+    /**
+     * Configures all the settings of each module.
+     */
+    private void configModules() {
+        for (int i = 0; i < 4; i++) {
+            modules[i] = new SwerveModule();
+            modules[i].setMotorPorts(Ports.SwerveDrive.motorPorts[i]);
+            modules[i].setAnglePID(Constants.SwerveDrive.anglePID[i]);
+            modules[i].setDrivePID(Constants.SwerveDrive.drivePID[i]);
+            modules[i].configInverted(Ports.SwerveDrive.motor_INVERTED[i]);
+            modules[i].configSensorPhase(Ports.SwerveDrive.motor_SENSOR_PHASE[i]);
+            modules[i].setZeroPosition(ZERO_POSITIONS[i]);
+        }
+    }
+
+    /**
+     * Terminates all movement of the swerve drive.
+     */
+    public void terminate() {
+        for (int i = 0; i < 4; i++) {
+            modules[i].terminate();
+        }
```

Comment:
Use the more-concise "for-each" loop:
```suggestion
        for (SwerveModule module : modules) {
            module.terminate();
        }
```